msprime/species_tree.py

```diff
@@ -0,0 +1,834 @@
+    PopulationConfiguration. However, these instances will be sorted
+    according to the alphabetical order of the corresponding species
+    names, allowing the user to link species from the species tree with
+    populations in msprime.
+
+    The sample size is not used except that it is stored in the
+    resulting instances of PopulationConfiguration.
+
+    Introgression events, as stored in the extended Newick format by
+    PhyloNet and possibly other programs, are not parsed and therefore
+    not used.
+    Rate matrices of continuous migration between co-existing lineages
+    can not be stored in (extended) Newick or Nexus format and are
+    therefore also not used here. However, these can be added to the
+    model after population configurations and demographic events are
+    defined with this function.
+    """
+
+    # Make sure a filename is specified.
+    err = 'A filename must be specified when calling function parse_species_tree().'
+    assert filename, err
+
+    # Make sure that branch length units are either "myr", "yr", or "gen".
+    allowed_branch_lenth_units = ["myr", "yr", "gen"]
+    if branch_length_units not in allowed_branch_lenth_units:
+        err = 'The specified units for branch lengths ('
+        err += '"{}") are not accepted. '.format(branch_length_units)
+        err += 'Accepted units are "myr" (millions of years), "yr" (years), '
+        err += 'and "gen" (generations).'
+        raise ValueError(err)
+
+    # Make sure that the sample size is either None or non-negative.
+    if sample_size is not None and sample_size < 0:
+        raise ValueError("Sample size must be >= 0")
+
+    # Make sure that the population size is either None or positive.
+    if Ne is not None and Ne <= 0:
+        raise ValueError("Population size Ne must be > 0")
+
+    # Make sure that the generation time is either None or positive.
+    if generation_time is not None and generation_time <= 0:
+        raise ValueError("Generation time must be > 0")
+
+    # Make sure that the generation time is specified if and only if
+    # branch lengths are not in units of generations.
+    if branch_length_units == "gen":
+        err = 'With branch lengths in units of generations ("gen"), generation_time '
+        err += 'should not be specified.'
+        assert generation_time is None, err
+    else:
+        err = 'With branch lengths in units of '
+        err += '"{}", generation_time must be specified.'.format(branch_length_units)
+        assert generation_time, err
+
+    # Get the number of generations per branch length unit.
+    if branch_length_units == "gen":
+        generations_per_branch_length_unit = 1
+    elif branch_length_units == "myr":
+        generations_per_branch_length_unit = 1000000/generation_time
+    else:
+        generations_per_branch_length_unit = 1/generation_time
+
+    # Read the input file.
+    tree_string = None
+    starbeast_format = False
```

Comment:
I don't think we should try to detect the format --- it's much simpler and more robust to make user do it (they'll know the format, presumably). 

I think we should split this into two functions: ``parse_species_tree`` and ``parse_starbeast``. That would simplify the parameterisation (we don't have to explain that one parameter is expected if it's in one format but not if it's in the other, etc), and make the code more readable.

Reply:
I now split the code into these two functions.